### PR TITLE
[fix][broker] fix npe when doCacheEviction

### PR DESCRIPTION
**Pull request by lordcheng10**
### Motivation
fix npe when doCacheEviction:
image

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `no-need-doc` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

**Comment by HQebupt** [comment]
LGTM

**Review by HQebupt — APPROVED** [review]

**Review by eolivelli — COMMENTED** [review]
Can we add a test case?

**Review by mattisonchao — APPROVED** [review]
LGTM +1.
Could we make `getEarlierReadPositionForActiveCursors` and `getEarlierMarkDeletedPositionForActiveCursors` return `Optional` to avoid potential NPE?

**Review by Jason918 — APPROVED** [review]
LGTM

**Review by 315157973 — APPROVED** [review]
Please address eolivelli‘s comment

**Review by Technoboy- — APPROVED** [review]

**Comment by lordcheng10** [comment]
> test cas

@eolivelli It is difficult to use the test case to simulate the scene where npe is generated.

**Comment by lordcheng10** [comment]
> Can we add a test case?

Fixed! add a test  method  testDoCacheEviction()  ,PTAL,thanks! @eolivelli

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java [hunk 1]
```diff
@@ -291,6 +291,68 @@ public void acknowledge1() throws Exception {
...
+                                ManagedLedger ledger = pair.getLeft();
+                                ManagedCursor cursor = pair.getRight();
+                                if (((ManagedLedgerImpl) ledger).getCacheSize() != message1.getBytes(Encoding).length) {
+                                    result.complete(false);
+                                    return;
+                                }
+
+                                ledger1.getActiveCursors().removeCursor(cursor.getName());
+                                try {
+                                    ((ManagedLedgerImpl) ledger).doCacheEviction(
```
Comment (eolivelli) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java:324]: we are not testing that this method is actually doing anything useful, this test only verifies that there is no runtime error.

can you please add some assertion that tells that the code works well even if `evictionPos` is null ?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java (hunk 1 above)
Comment (lordcheng10) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java:324]: OK

Comment (lordcheng10) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java:325]: Fixed. PTAL,thanks! @eolivelli

**Comment by lordcheng10** [comment]
The command to re-trigger CI seems to be invalid: /pulsarbot run-failure-checks , I have entered multiple times, but it still does not trigger.
image

**Review by eolivelli — APPROVED** [review]
LGTM

nice work

**Comment by lordcheng10** [comment]
@aloyszhang PTAL,thank！

**Review by aloyszhang — APPROVED** [review]

**Comment by mattisonchao** [comment]
Hi, @lordcheng10 
It looks like do not affect branch 2.9. Could you help to confirm it?

**Comment by Jason918** [comment]
This does not affect branch-2.10, remove 'release/2.10.2'

**Pull request merged**